### PR TITLE
Feat: Loading Hints

### PR DESCRIPTION
**Pull request by mOctave**
**Feature:** This PR adds the feature suggested in this issue (closes) #7583.

## Feature Details
The game now pulls loading hints from a file called `loading hints.txt` (the phrase `"loading hints"`) and displays them during startup (currently not randomizing properly) and on the main menu, choosing a new one every time the menu screen is opened.

## UI Screenshots

Screenshot 2023-02-02 at 8 28 42 PM

## Usage Examples

Five new phrases are defined here: `"loading hints"`, `"loading hints (advice)"`, `"loading hints (community)"`, `"loading hints (controls)"`, and `"loading hints (trivia)"`. If, for example, a plugin author wanted to add a link to their webpage, some advice about their plugin, and some trivia, they could do so with:
```
phrase "loading hints (advice)"
	word
		"Update LoadingHintPlugin often to get new loading hints!"
		"Watch out for the wild aliens that control the Republic."
		"Never eat astronomical daisies."

phrase "loading hints (community)"
	word
		"There's a website for LoadingHintPlugin at www.example.com."

phrase "loading hints (trivia)"
	word
		"Tomorrow, we're celebrating LoadingHintPlugin's one-day anniversary!"
```

## Testing Done
Loading Endless Sky a whole bunch of times.

### Automated Tests Added
{{describe the automated tests you added (using the unit-test framework, integration-test framework or otherwise) to reduce the risk of regressions in the future. "N/A" if this PR is not for an in-game feature that needs to remain functional in the future. }}

## Performance Impact
N/A

## Not Ready to Open Yet Because:

- [ ] Random hint selection isn't working
- [x] Text replacement isn't working
- [x] Switch to using phrases
- [x] It might be possible to define this in interfaces.txt
- [x] Centre
- [ ] New box to put the hint in?

### CODE REVIEW

File: source/LoadingHints.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <string>
+#include <vector>
+
+class LoadingHints {
+public:
```
Comment (RisingLeaf) [inline source/LoadingHints.h:20]:
```suggestion
namespace {
```

File: source/GameLoadingPanel.cpp [hunk 2]
```diff
@@ -36,7 +44,10 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "opengl.h"
 
-
+namespace {
+	std::vector<std::string> loadingHintArray = LoadingHints().loadingHintArray;
+	std::string hint = loadingHintArray[Random::Int(sizeof(loadingHintArray))];
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:49]:
```suggestion
	std::string hint;
```

File: source/GameLoadingPanel.cpp [hunk 3]
```diff
@@ -108,4 +119,12 @@ void GameLoadingPanel::Draw()
 		a += da;
 	}
 	PointerShader::Unbind();
+
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:122]:
```suggestion

	hint = loadingHintArray[Random::Int(sizeof(loadingHintArray))];
```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/LoadingHints.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		"Sell and change outfits on multiple ships at once by selecting them with shift or control-clicking.",
+		"Board ships to earn money and build your fleet.",
+		"Check your piracy threat often on the info screen, and keep it as low as possible.",
+		"There are over 97 billion humans in the galaxy.",
+		"The inhabited planet with the lowest human population is New Sahara, which is home to only 3000 people.",
+		"Every manufacturer's ships have their own strengths and weaknesses.",
+		"You can use ESLauncher2 to install and play Endless Sky more easily.",
+		""
+	};
+};
```
Comment (RisingLeaf) [inline source/LoadingHints.h:51]:
```suggestion
}

```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/GameLoadingPanel.cpp (hunk 3 above)
Comment (mOctave) [reply source/GameLoadingPanel.cpp:125]: Doing it this way leads to the hint being re-chosen every time the loading screen is drawn, but I've moved it to GameLoadingPanel::GameLoadingPanel(), where I think it works okay.

**Comment by bene-dictator** [comment]
A few more tips:
- "Meteor missiles don't track ships that run cold."
- "If you're low on credits, the Spaceport can be a useful place to find higher-paying missions." 
- "Using too many outfit expansions may cause your cooling outfits to stop working."
- "You can rename your ships in your fleet information panel by pressing 'I' and selecting a ship."
- "You can reorder your ships' weapons by pressing 'I', selecting a ship, and dragging the weapons around."
- "Missiles are powerful alternatives to guns if you want to destroy ships quick and fast."
- "Loans can be very useful for starting a fleet quickly."
- "There are often more to uninhabited regions of space than you think."
- "Pirates sell some special outfits that civilian outfitters don't stock."
- "Afterburners are a great way to catch up to someone or to flee from them."
- "Karate is never the best way to greet someone."

There's already some tips that say the same things, but I want to hammer those in a bit more. I don't know if it would work here, but try changing some of the tips that tell you to press the default key to showing the keys that the player has mapped (for example, 'I' to `<View player info>`)

File: source/GameLoadingPanel.cpp (hunk 3 above)
Comment (RisingLeaf) [reply source/GameLoadingPanel.cpp:125]: oh, I wanted to place it there, got the wrong line

**Comment by Saugia** [comment]
Personally, I find the home screen to be a very odd place for loading hints. From a game design perspective, it really doesn't make sense for Endless Sky. It'd make more sense if we had scenes and loading time in-between things, but cluttering the beginning screen doesn't feel right to me. For a game like Endless Sky, that barely has load times and has plenty of downtime during missions or landing, it's really easy to just google or learn these little hints on your own.

**Comment by bene-dictator** [comment]
Yeah, it is a little awkward. It still takes about 5 - 10 seconds for me to load Endless Sky with a decent computer though. I think we should have the player search for tips for the game as few times as possible and have the game show the tips... but through the introduction, the story, or characters' dialogue. In-game tips are only a little better than googling.

**Comment by Zitchas** [comment]
Two thoughts on this:
1) Yes! Definitely, I think having tips on the main menu is a great thing. Both hint-tips like the ones you have and Bene-dectator suggested; and more specific UI tips. Ex. "In the outfitter hit "u" to uninstall" sorts of things. Every somewhat obscure key, shortcut, or other command that isn't explicitly *stated* in game should get a stand-alone tip to go on the screen. In an ideal world, players should *never,* **ever**, have to go anywhere outside the game to find out any of the controls, interface, commands, or methodology for using the game and its basic systems. 

2) I don't think the bottom left corner is a great place for this. It needs to be a little bit more in the field of view. Bigger, too. I'd actually be inclined to make a text box/window in the same style as the other two on the main screen, except really wide but very short, and put it straight across the center/bottom, right over top of the planet.

**Comment by Zitchas** [comment]
A question / suggestion: 

Would it be possible to have it pull the hints from a data file instead of having them baked into the source? It'd be great if they were more readily accessible - both for ease of adding new ones and making it easier for players to access the full list if they want. (and then plug-ins could add more, too.)

**Comment by mOctave** [comment]
> For a game like Endless Sky, that barely has load times and has plenty of downtime during missions or landing, it's really easy to just google or learn these little hints on your own.

For most players with good computers and not a lot of plugins, this is true, but anyone who develops their own plugins or plays a large variety of other people's plugins probably frequently gets at least a 5-10 second load time. Furthermore, there are some really odd shortcuts in Endless Sky that you would never know about without reading the hard-to-find player manual or being on the Discord server.

> Regardless, the bottom left corner is not a good place for this UI element at all. It's far too out of the way for something as important as a tip, and is further away from the rest of the UI than anything else. From a design perspective, it completely ruins the balance of the UI.

> I don't think the bottom left corner is a great place for this. It needs to be a little bit more in the field of view. Bigger, too. I'd actually be inclined to make a text box/window in the same style as the other two on the main screen, except really wide but very short, and put it straight across the center/bottom, right over top of the planet.

Agreed. I'll see if I can centre it while adding it to interfaces.txt.

> Would it be possible to have it pull the hints from a data file instead of having them baked into the source?

As far as I know, no. If it turns out that it is, I'll definitely see about switching to it, since I agree that a datafile-based system would be much nicer.

> I don't know if it would work here, but try changing some of the tips that tell you to press the default key to showing the keys that the player has mapped (for example, 'I' to <View player info>)

I forgot about custom keybindings... I'll see if that's possible, and if it is, I'll try to add it.

**Comment by bene-dictator** [comment]
> I forgot about custom keybindings... I'll see if that's possible, and if it is, I'll try to add it.

If the list of hints becomes a data file, it should be possible because I got it from help.txt

**Comment by mOctave** [comment]
> If the list of hints becomes a data file, it should be possible because I got it from help.txt

It might not be. As far as I know, interfaces.txt _might_ be loaded early, but all the rest of the datafiles are loaded after the hint gets displayed. This might also be a reason why it _wouldn't_ work in a datafile.

**Comment by Zitchas** [comment]
Maybe a combined method? Have as many as possible hardcoded (so these would be the ones displayed during startup), but have it load a data file during the regular loading process. Then whenever the user goes to the main screen for some reason, it can rotate through both the regular hardcoded ones and the ones loaded from the data file.

**Comment by mOctave** [comment]
> Maybe a combined method? Have as many as possible hardcoded (so these would be the ones displayed during startup), but have it load a data file during the regular loading process. Then whenever the user goes to the main screen for some reason, it can rotate through both the regular hardcoded ones and the ones loaded from the data file.

That might work!

**Comment by 1010todd** [comment]
> Personally, I find the home screen to be a very odd place for loading hints. From a game design perspective, it really doesn't make sense for Endless Sky. It'd make more sense if we had scenes and loading time in-between things, but cluttering the beginning screen doesn't feel right to me. For a game like Endless Sky, that barely has load times and has plenty of downtime during missions or landing, it's really easy to just google or learn these little hints on your own.
> 
> Regardless, the bottom left corner is not a good place for this UI element at all. It's far too out of the way for something as important as a tip, and is further away from the rest of the UI than anything else. From a design perspective, it completely ruins the balance of the UI.

Home screen is the the only out-of-character place in the entire game, it would be much weirder to be anywhere else in the game. Maybe only preference/setting page would be another ooc place but it would make even less sense to put tips there as typically you only go there once in a while resulting in much fewer hints shown to players and generally became less useful. Other fairly popular games like Wesnoth and Starsector also put hints and tips on home screen, it'll be fine. Telling people to go out of the game and google for hints sounds like a much worse design, I think a game should be able to show the player each features within the game itself and relying on an official manual at worst.

Though, I do agree bottom left corner is not a good place. Maybe I would try putting it in the middle under the compass / loading circle but it goes over the planet which I'm not sure how we feel about that.

**Comment by bene-dictator** [comment]
It might look less out of place if it has a grey box around it like all the other text on the screen, and if it was placed in the centre, below the planet.

**Comment by Amazinite** [comment]
> Random hint selection isn't working
> I'm guessing the way I'm selecting the hint could be seriously optimized

Make use of the [Phrase](https://github.com/endless-sky/endless-sky/blob/master/source/Phrase.cpp) class. This is already the class we use for these types of things. It's used for ship names, ship hails, spaceport news, and more. By using a phrase, you allow the creation of new hints to be done in the game data instead of hints being hard coded. Phrase also already randomly selects a sentence from among its possible sentences.

The main problem with this is that the phrase would need to be loaded first, so you can't really have a "loading hint" before the game has actually loaded. Given how quickly the game typically loads though (assuming you don't have a ton of plugins), I don't think it's critical that we have these ready during loading anyway.

> It might be possible to define this in interfaces.txt
> Centre

This should be defined by an interface, which can define a text `box` with a certain location and size. The menu background and main menu are already existing interfaces that you could potentially use. 

> Performance impact not measured yet

This isn't performance-critical code, therefore you can just say N/A.

**Comment by RisingLeaf** [comment]
> > Random hint selection isn't working
> > I'm guessing the way I'm selecting the hint could be seriously optimized
> 
> Make use of the [Phrase](https://github.com/endless-sky/endless-sky/blob/master/source/Phrase.cpp) class. This is already the class we use for these types of things. It's used for ship names, ship hails, spaceport news, and more. By using a phrase, you allow the creation of new hints to be done in the game data instead of hints being hard coded. Phrase also already randomly selects a sentence from among its possible sentences.

They dont want hints to be loaded, because they say, hints should be displayed while things get loaded.

**Comment by mOctave** [comment]
> This isn't performance-critical code, therefore you can just say N/A

Good to know.

> This should be defined by an interface, which can define a text box with a certain location and size. The menu background and main menu are already existing interfaces that you could potentially use.

I'll see if I can get this working within the next half-hour or so!

> Although maybe at that point they'd be more like main menu hints than loading hints.

As RisingLeaf said, I'd like the hints to be loaded earlier so that they make loading the game a lot less unbearable if you have a lot of plugins (Hecter's "every plugin launch" comes to mind as an extreme case), some broken plugins (which spend a lot of time logging everything they fail to do and loading things very slowly, a situation I often encounter), or just a slow computer. I said this on Discord, and I probably should have said it here as well – I will when this is ready for review:

__My Reasons for Doing This__
1. Provide hints for new players or insights for casual players.
2. Add flavour to the game.
3. Give you something to look at if you're loading a lot of plugins.

On a different note, I'm not quite sure what's causing the Ubuntu tests to fail despite checking the details. I'd appreciate some insight on this by someone who is better at reading error logs than I am.

File: source/MenuPanel.cpp [hunk 5]
```diff
@@ -100,6 +102,7 @@ void MenuPanel::Draw()
 	glClear(GL_COLOR_BUFFER_BIT);
 	GameData::Background().Draw(Point(), Point());
 
+	// Load the player information.
```
Comment (RisingLeaf) [inline source/MenuPanel.cpp:105]:
```suggestion
	// Check if a player is loaded.
```

File: source/LoadingHints.h (hunk 4 above)
Comment (RisingLeaf) [reply source/LoadingHints.h:53]: I dont think we need a class here, a namespace should be enough

File: source/GameLoadingPanel.h [hunk 6]
```diff
@@ -50,6 +49,4 @@ class GameLoadingPanel final : public Panel {
 	int progress = 0;
 };
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.h:51]:
```suggestion



```

File: source/GameLoadingPanel.h [hunk 7]
```diff
@@ -36,6 +34,7 @@ class GameLoadingPanel final : public Panel {
 	void Step() final;
 	void Draw() final;
 
+	static std::string GetHint();
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.h:37]:
```suggestion
	static std::string GetHint();

```

File: source/GameLoadingPanel.cpp [hunk 8]
```diff
@@ -108,4 +122,14 @@ void GameLoadingPanel::Draw()
...
+	// Draw the loading hint.
+	Information info;
+	info.SetString("hint", GameLoadingPanel::GetHint());
+	GameData::Interfaces().Get("menu background")->Draw(info, this);
+}
+
+std::string GameLoadingPanel::GetHint()
+{
+	return hint;
 }
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:135]:
```suggestion
}

```

File: source/GameLoadingPanel.cpp [hunk 9]
```diff
@@ -108,4 +122,14 @@ void GameLoadingPanel::Draw()
 		a += da;
 	}
 	PointerShader::Unbind();
+
+	// Draw the loading hint.
+	Information info;
+	info.SetString("hint", GameLoadingPanel::GetHint());
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:128]:
```suggestion
	info.SetString("hint", GetHint());
```

File: source/GameLoadingPanel.cpp [hunk 10]
```diff
@@ -36,14 +44,20 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
 GameLoadingPanel::GameLoadingPanel(PlayerInfo &player, const Conversation &conversation,
 	UI &gamePanels, bool &finishedLoading)
 	: player(player), conversation(conversation), gamePanels(gamePanels),
 		finishedLoading(finishedLoading), ANGLE_OFFSET(360. / MAX_TICKS)
 {
 	SetIsFullScreen(true);
+
+	// Choose the loading hint.
+	std::vector<std::string> loadingHintArray = LoadingHints().loadingHintArray;
+	hint = loadingHintArray[Random::Int(sizeof(loadingHintArray))];
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:60]:
```suggestion
	hint = LoadingHints().loadingHintArray[Random::Int(sizeof(LoadingHints().loadingHintArray))];
```

no need for a local copy

File: source/GameLoadingPanel.cpp [hunk 11]
```diff
@@ -36,14 +44,20 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "opengl.h"
 
-
+namespace {
+	std::string hint;
+}
```
Comment (RisingLeaf) [inline source/GameLoadingPanel.cpp:49]:
```suggestion
}


```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

**Comment by mOctave** [comment]
Currently, this is waiting on https://github.com/endless-sky/endless-sky/pull/8054 and `Phrase::GetAll()`. After that gets merged, I'll get to work on fixing random hint selection.

**Comment by Koranir** [comment]
I feel that the text would be a bit more readable if it went a little underneath the planet, so there is some more contrast.

**Comment by mOctave** [comment]
> I feel that the text would be a bit more readable if it went a little underneath the planet, so there is some more contrast.

Right now, the text placement is fairly temporary, although I agree that it would probably work best under the planet in the end.

**Comment by Amazinite** [comment]
I'm not really sold on the idea of implementing loading hints. Loading hints is a very cliché thing that lots of games do, but honestly I don't feel like they're every really that helpful. Per the original issue that this was spawned from. if the underlying problem is that there exists things in the game that the player can't know about without searching outside of the game, then I don't think that loading hints are our best way of solving that, particularly because of how one of the major points being made in the comments of this PR about adding loading hints isn't to help players get information they might not know, but instead give people with a lot of plugins something to look at while the game is loading. But wouldn't you think that someone with a lot of plugins is probably also someone with lots of knowledge about the game already? And aren't the people who need the hints the most also probably people who are playing vanilla and will therefore get the quickest loading times?

For the underlying problem of providing more information inside the game, I think should instead find better ways to make sure that information is known, not ways like loading hints where whether we ever end up actually giving the player useful into is entirely random. I'd suggest looking to an issue like #3663 as a means of providing players within information in a centralized source where they're guaranteed to see it if they go looking for it instead of it randomly being shown to them.

**Pull request closed without merging**